### PR TITLE
OBJ-233 Remove objection id from web session after submission

### PR DESCRIPTION
**Pull request by mwestacott**
Also correctly handle any errors when submitting objections

### CODE REVIEW

File: src/controllers/check.your.answers.controller.ts [hunk 1]
```diff
@@ -39,8 +39,12 @@ export const get = async (req: Request, res: Response, next: NextFunction) => {
...
 
-export const post = async (req: Request, res: Response) => {
-  await submitObjection(req.session as Session);
-
-  return res.redirect(OBJECTIONS_CONFIRMATION);
+export const post = async (req: Request, res: Response, next: NextFunction) => {
+  try {
+    await submitObjection(req.session as Session);
+    return res.redirect(OBJECTIONS_CONFIRMATION);
+  } catch (e) {
```
Comment (mwestacott) [inline src/controllers/check.your.answers.controller.ts:46]: This stops the web app timing out when there’s an issue submitting the objection. If the REST call fails for any reason the ‘Sorry, there is a problem with the service’ error screen will display promptly. @markpit this changes behaviour for the JS off - double-click scenario but I think it’s required as we then correctly handle all the other potential error scenarios.

File: src/controllers/confirmation.controller.ts [hunk 2]
```diff
@@ -2,13 +2,17 @@ import { Session } from "ch-node-session-handler";
...
+import { deleteFromObjectionSession, retrieveFromObjectionSession, retrieveUserEmailFromSession } from "../services/objection.session.service";
 
 export const get = (req: Request, res: Response, next: NextFunction) => {
   const session: Session | undefined  = req.session;
   if (session) {
     const email: string = retrieveUserEmailFromSession(session);
     const objectionId: string = retrieveFromObjectionSession(session, SESSION_OBJECTION_ID);
+
+    // Objection has been submitted and should no longer be referenced by this web session, so remove the objection id from the session
+    deleteFromObjectionSession(session, SESSION_OBJECTION_ID);
```
Comment (mwestacott) [inline src/controllers/confirmation.controller.ts:14]: This addresses the additional issue highlighted in Dale’s final comment on OBJ-233 - navigating around the web app by changing URLs and being able to change the Objection Mongo DB data after it’s been submitted.

**Comment by markpit** [comment]
looks ok to me

**Review by ICoxCH — APPROVED** [review]

File: test/controller/confirmation.controller.spec.unit.ts [hunk 3]
```diff
@@ -60,4 +64,40 @@ describe("confirmation screen tests", () => {
...
+    expect(response.text).toContain(OBJECTION_ID);
+    expect(response.text).toContain(email);
+  });
+
+  it("should land on error screen if no session is available", async () => {
+    mockRetrieveUserEmailFromSession.mockReturnValueOnce(email);
+    mockRetrieveFromObjectionSession.mockReturnValueOnce(OBJECTION_ID);
+
+    mockSessionMiddleware.mockReset();
+    mockSessionMiddleware.mockImplementation((next: NextFunction) => {
```
Comment (markpit) [inline test/controller/confirmation.controller.spec.unit.ts:91]: I'd go with mockImplementationOnce here, i'm not sure what order tests will run in.

**Review by markpit — APPROVED** [review]

**Comment by mwestacott** [comment]
Won't merge until https://github.com/companieshouse/strike-off-objections-api/pull/93 is complete and approved.

**Pull request merged**